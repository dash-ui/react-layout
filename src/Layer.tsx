import * as React from 'react'
import clsx from 'clsx'
import css from 'minify-css.macro'
import {Frame} from './Frame'
import {useLayout} from './Layout'
import type {FrameProps} from './Frame'
import type {MediaQueryProp, MediaQueries} from './Layout'

/**
 * A component that makes its items layer on top of each other:
 *   ___
 *  | ☐ |
 *   ‾‾‾
 */
export const Layer = React.forwardRef<any, LayerProps>((props, ref) => (
  <Frame ref={ref} position="relative" {...props} />
))

export const LayerItem = React.forwardRef<any, LayerItemProps>(
  ({className, placement, offset, z, ...props}, ref) => {
    const prop = useLayout().mq.prop

    return (
      <Frame
        ref={ref}
        position="absolute"
        className={clsx(
          className,
          prop(placementStyle(offset), placement),
          prop(zStyle, z)
        )}
        {...props}
      />
    )
  }
)

const placementStyle = (offsetProp: LayerItemProps['offset'] | undefined) => (
  value: Placements,
  queryName: keyof MediaQueries
) => {
  if (value === 'center') {
    return css`
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
    `
  }

  const offset = !offsetProp
    ? 0
    : typeof offsetProp === 'object'
    ? offsetProp[queryName]
    : offsetProp
  const lValue = value.toLowerCase()
  const yProp =
    lValue.indexOf('top') > -1
      ? 'top'
      : lValue.indexOf('bottom') > -1
      ? 'bottom'
      : void 0
  const xProp =
    lValue.indexOf('left') > -1
      ? 'left'
      : lValue.indexOf('right') > -1
      ? 'right'
      : void 0

  const styles: Record<string, string | number> = {}

  if (yProp) styles[yProp] = offset
  if (xProp) styles[xProp] = offset
  if (value === 'left' || value === 'right') {
    styles.top = '50%'
    styles.transform = 'translateY(-50%)'
  }
  if (value === 'bottom' || value === 'top') {
    styles.left = '50%'
    styles.transform = 'translateX(-50%)'
  }

  return styles
}

const zStyle = (value: number) => ({zIndex: value})

export interface LayerProps extends FrameProps {}

type Placements =
  | 'top'
  | 'right'
  | 'bottom'
  | 'left'
  | 'center'
  | 'topLeft'
  | 'topRight'
  | 'bottomRight'
  | 'bottomLeft'

export interface LayerItemProps extends FrameProps {
  position?: undefined
  offset?: MediaQueryProp<number | string>
  placement: MediaQueryProp<Placements>
  z?: MediaQueryProp<number>
}
