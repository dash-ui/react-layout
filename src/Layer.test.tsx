/* jest */
import * as React from 'react'
import {render} from '@testing-library/react'
import {Layer, LayerItem} from './Layer'
import {renderMq, mediaQueries} from 'test-utils'

describe('<Layer>', () => {
  it('should apply default styles', () => {
    const {getByTestId} = render(<Layer data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule('position', 'relative')
  })
})

describe('<LayerItem> without media queries', () => {
  it('applies default styles for each position', () => {
    const placements = [
      'center',
      'topLeft',
      'top',
      'topRight',
      'right',
      'bottomRight',
      'bottom',
      'bottomLeft',
      'left',
    ]
    for (const placement of placements) {
      const {asFragment} = render(<LayerItem placement={placement as any} />)
      expect(asFragment()).toMatchSnapshot(placement)
    }
  })

  it('applies styles with offsets for each position', () => {
    const placements = [
      'topLeft',
      'top',
      'topRight',
      'right',
      'bottomRight',
      'bottom',
      'bottomLeft',
      'left',
    ]
    for (const placement of placements) {
      const {asFragment} = render(
        <LayerItem placement={placement as any} offset={10} />
      )
      expect(asFragment()).toMatchSnapshot(placement)
    }
  })

  it('applies the "z" prop', () => {
    const {getByTestId} = render(
      <LayerItem placement="top" z={1} data-testid="el" />
    )
    expect(getByTestId('el')).toHaveStyleRule('z-index', '1')
  })
})

describe('<LayerItem> with media queries', () => {
  it('applies default styles for each position', () => {
    const placements = [
      'center',
      'topLeft',
      'top',
      'topRight',
      'right',
      'bottomRight',
      'bottom',
      'bottomLeft',
      'left',
    ]
    for (const placement of placements) {
      const {asFragment} = renderMq(
        <LayerItem placement={{phone: placement as any}} />
      )
      expect(asFragment()).toMatchSnapshot(placement)
    }
  })

  it('applies styles with offsets for each position', () => {
    const placements = [
      'topLeft',
      'top',
      'topRight',
      'right',
      'bottomRight',
      'bottom',
      'bottomLeft',
      'left',
    ]
    for (const placement of placements) {
      const {asFragment} = renderMq(
        <LayerItem placement={{phone: placement as any}} offset={10} />
      )
      expect(asFragment()).toMatchSnapshot(placement)
    }
  })

  it('applies the "z" prop', () => {
    const {getByTestId} = renderMq(
      <LayerItem placement="top" z={{phone: 1}} data-testid="el" />
    )

    expect(getByTestId('el')).toHaveStyleRule('z-index', '1', {
      media: mediaQueries.phone,
    })
  })
})
