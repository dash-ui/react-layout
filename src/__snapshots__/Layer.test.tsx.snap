// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LayerItem> with media queries applies default styles for each position: bottom 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    bottom: 0;
    left: 50%;
    -webkit-transform: translateX(-50%);
    -ms-transform: translateX(-50%);
    transform: translateX(-50%);
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies default styles for each position: bottomLeft 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    bottom: 0;
    left: 0;
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies default styles for each position: bottomRight 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    bottom: 0;
    right: 0;
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies default styles for each position: center 1`] = `
<DocumentFragment>
  @media only screen and (min-width:0em) {
  .-ui-0 {
    top: 50%;
    left: 50%;
    -webkit-transform: translate(-50%,-50%);
    -ms-transform: translate(-50%,-50%);
    transform: translate(-50%,-50%);
  }
}

.-ui-1 {
  position: absolute;
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies default styles for each position: left 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    left: 0;
    top: 50%;
    -webkit-transform: translateY(-50%);
    -ms-transform: translateY(-50%);
    transform: translateY(-50%);
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies default styles for each position: right 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    right: 0;
    top: 50%;
    -webkit-transform: translateY(-50%);
    -ms-transform: translateY(-50%);
    transform: translateY(-50%);
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies default styles for each position: top 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    top: 0;
    left: 50%;
    -webkit-transform: translateX(-50%);
    -ms-transform: translateX(-50%);
    transform: translateX(-50%);
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies default styles for each position: topLeft 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    top: 0;
    left: 0;
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies default styles for each position: topRight 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    top: 0;
    right: 0;
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies styles with offsets for each position: bottom 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    bottom: 10px;
    left: 50%;
    -webkit-transform: translateX(-50%);
    -ms-transform: translateX(-50%);
    transform: translateX(-50%);
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies styles with offsets for each position: bottomLeft 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    bottom: 10px;
    left: 10px;
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies styles with offsets for each position: bottomRight 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    bottom: 10px;
    right: 10px;
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies styles with offsets for each position: left 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    left: 10px;
    top: 50%;
    -webkit-transform: translateY(-50%);
    -ms-transform: translateY(-50%);
    transform: translateY(-50%);
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies styles with offsets for each position: right 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    right: 10px;
    top: 50%;
    -webkit-transform: translateY(-50%);
    -ms-transform: translateY(-50%);
    transform: translateY(-50%);
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies styles with offsets for each position: top 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    top: 10px;
    left: 50%;
    -webkit-transform: translateX(-50%);
    -ms-transform: translateX(-50%);
    transform: translateX(-50%);
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies styles with offsets for each position: topLeft 1`] = `
<DocumentFragment>
  @media only screen and (min-width:0em) {
  .-ui-0 {
    top: 10px;
    left: 10px;
  }
}

.-ui-1 {
  position: absolute;
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> with media queries applies styles with offsets for each position: topRight 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

@media only screen and (min-width:0em) {
  .-ui-0 {
    top: 10px;
    right: 10px;
  }
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies default styles for each position: bottom 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  bottom: 0;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies default styles for each position: bottomLeft 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  bottom: 0;
  left: 0;
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies default styles for each position: bottomRight 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  bottom: 0;
  right: 0;
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies default styles for each position: center 1`] = `
<DocumentFragment>
  .-ui-0 {
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.-ui-1 {
  position: absolute;
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies default styles for each position: left 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  left: 0;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies default styles for each position: right 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  right: 0;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies default styles for each position: top 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  top: 0;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies default styles for each position: topLeft 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  top: 0;
  left: 0;
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies default styles for each position: topRight 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  top: 0;
  right: 0;
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies styles with offsets for each position: bottom 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  bottom: 10px;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies styles with offsets for each position: bottomLeft 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  bottom: 10px;
  left: 10px;
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies styles with offsets for each position: bottomRight 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  bottom: 10px;
  right: 10px;
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies styles with offsets for each position: left 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  left: 10px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies styles with offsets for each position: right 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  right: 10px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies styles with offsets for each position: top 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  top: 10px;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies styles with offsets for each position: topLeft 1`] = `
<DocumentFragment>
  .-ui-0 {
  top: 10px;
  left: 10px;
}

.-ui-1 {
  position: absolute;
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;

exports[`<LayerItem> without media queries applies styles with offsets for each position: topRight 1`] = `
<DocumentFragment>
  .-ui-1 {
  position: absolute;
}

.-ui-0 {
  top: 10px;
  right: 10px;
}

<div
    class="-ui-0 -ui-1"
  />
</DocumentFragment>
`;
